### DISCUSSION TITLE
configuring relationship foreign() , one-to-many / many-to-one, and using with_parent()

### DISCUSSION THREAD

**Discussion by wwoods** [sqlalchemy/sqlalchemy#6859] [Usage Questions]
### Describe the bug

I'm trying to build a view of SQLAlchemy models over existing SQLAlchemy models, which leverage some intermediate processing. This can be thought of as a fused view over a graph. The attached MWE demonstrates this schema.

When using `sqlalchemy.orm.with_parent` in this scheme, to avoid a `lazy="dynamic"` relationship, I'm getting the full table rather than rows filtered correctly on the parent object. Printing the SQL shows that the entity specified in `with_parent` is not integrated into the query. Seems like a bug.

### To Reproduce

```python
import sqlalchemy as sa
import sqlalchemy.orm

Base = sa.orm.declarative_base()

class Obj(Base):
    __tablename__ = 'obj'
    id = sa.Column(sa.Integer, primary_key=True)

class Link(Base):
    def __repr__(self):
        return f'<Link {self.id} / {self.id_src} -> {self.id_dst}>'
    __tablename__ = 'link'
    id = sa.Column(sa.Integer, primary_key=True)
    id_src = sa.Column(sa.Integer, sa.ForeignKey('obj.id'))
    id_dst = sa.Column(sa.Integer, sa.ForeignKey('obj.id'))

class Remap(Base):
    __tablename__ = 'remap'
    id_low = sa.Column(sa.Integer)
    id_high = sa.Column(sa.Integer, autoincrement=False, primary_key=True)

_obj_lowest = sa.select(Remap).where(Remap.id_low == Obj.id).exists()
_obj_query = sa.select(Obj).where(_obj_lowest).subquery()
class FusedObj(Base):
    def __repr__(self):
        return f'<FusedObj {self.id}>'
    __table__ = _obj_query

_F1 = sa.orm.aliased(Remap)
_F2 = sa.orm.aliased(Remap)
_link_query = (
        sa.select(Link.id, _F1.id_low.label('id_src'), _F2.id_low.label('id_dst'))
        .select_from(Link)
        .join(_F1, Link.id_src == _F1.id_high)
        .join(_F2, Link.id_dst == _F2.id_high)
        ).subquery()
class FusedLink(Base):
    def __repr__(self):
        return f'<FusedLink {self.id} / {self.id_src} -> {self.id_dst}>'
    __table__ = _link_query

    src = sa.orm.relationship('FusedObj',
            backref=sa.orm.backref('link_as_src', lazy='raise'),
            primaryjoin='remote(FusedLink.id_src) == foreign(FusedObj.id)',
            viewonly=True)


def main():
    engine = sa.create_engine(f'sqlite://', future=True)
    Base.metadata.create_all(engine)

    sessionmaker = sa.orm.sessionmaker(engine)

    with sessionmaker.begin() as sess:
        for i in range(1, 6):
            sess.execute(sa.insert(Obj).values(id=i))
        for i in range(1, 5):
            sess.execute(sa.insert(Link).values(id_src=i, id_dst=i+1))
        for i in range(2, 4):
            sess.execute(sa.insert(Remap).values(id_low=1, id_high=i))
        for i in range(4, 6):
            sess.execute(sa.insert(Remap).values(id_low=i, id_high=i))

        fused_1 = sess.execute(sa.select(FusedObj).where(FusedObj.id == 1)).scalar()
        # Trouble line!!!
        query = sa.select(FusedLink).where(sa.orm.with_parent(fused_1,
                FusedLink.src))
        print(str(query))
        rel = sess.execute(query).all()
        print('Should only be 2 FusedLink:')
        print(rel)

        print('Underlying objects:')
        print(sess.execute(sa.select(FusedObj)).all())
        print(sess.execute(sa.select(Link)).all())
        print(sess.execute(sa.select(FusedLink)).all())


if __name__ == '__main__':
    main()
```

### Error

No object identifier in produced query:

```
SELECT anon_1.id, anon_1.id_src, anon_1.id_dst 
FROM (SELECT link.id AS id, remap_1.id_low AS id_src, remap_2.id_low AS id_dst 
FROM link JOIN remap AS remap_1 ON link.id_src = remap_1.id_high JOIN remap AS remap_2 ON link.id_dst = remap_2.id_high) AS anon_1, (SELECT obj.id AS id 
FROM obj 
WHERE EXISTS (SELECT remap.id_low, remap.id_high 
FROM remap 
WHERE remap.id_low = obj.id)) AS anon_2 
WHERE anon_1.id_src = anon_2.id
Should only be 2 FusedLink:
[(<FusedLink 2 / 1 -> 1>,), (<FusedLink 3 / 1 -> 4>,), (<FusedLink 4 / 4 -> 5>,)]
Underlying objects:
[(<FusedObj 1>,), (<FusedObj 4>,), (<FusedObj 5>,)]
[(<Link 1 / 1 -> 2>,), (<Link 2 / 2 -> 3>,), (<Link 3 / 3 -> 4>,), (<Link 4 / 4 -> 5>,)]
[(<FusedLink 2 / 1 -> 1>,), (<FusedLink 3 / 1 -> 4>,), (<FusedLink 4 / 4 -> 5>,)]
```

### Versions

- OS: Manjaro linux
- Python: 3.8.5
- SQLAlchemy: 1.4.22
- Database: sqlite + postgres
- DBAPI:

**Comment by zzzeek** [comment]
the immediate issue is first that the relationship is not configured correctly:

```py
class FusedLink(Base):
    __table__ = _link_query

    src = sa.orm.relationship('FusedObj',
            backref=sa.orm.backref('link_as_src', lazy='raise'),
            primaryjoin='remote(FusedLink.id_src) == foreign(FusedObj.id)',
            viewonly=True)
```

`remote(FusedLink.id_src)` is incorrect, because by definition, the remote side of this relationship is `FusedObj`, it's the "thing this relationship points towards".   There is never a reason to use `remote()` unless the relationship is self-referential, that is, points FusedLink to FusedLink.

from there I think `foreign()` here is `FusedLink.id_src`, not `FusedObj.id`.  `FusedObj.id` AFAICT is a unique primary key, and many `FusedLink.id_src` entries may refer to a single `FusedObj.id`; the output at the end is `[(<FusedLink 2 / 1 -> 1>,), (<FusedLink 3 / 1 -> 4>,), (<FusedLink 4 / 4 -> 5>,)]`, so that means both FusedLink(2) and FusedLink(3) point to `id_src = 1`, this is the foreign column:

```py
class FusedLink(Base):
    __table__ = _link_query

    src = sa.orm.relationship('FusedObj',
            backref=sa.orm.backref('link_as_src', lazy='raise'),
            primaryjoin='foreign(FusedLink.id_src) == FusedObj.id',
            viewonly=True)
```

So this makes FusedLink.src a many-to-one relationship.   

So we can query for this either as:

```py
        query = sa.select(FusedLink).where(
            FusedLink.src == fused_1
        )
```


or using the one-to-many side:

```py
        query = sa.select(FusedLink).where(
             sa.orm.with_parent(fused_1, FusedObj.link_as_src)
        )
```


both give the SQL:

```
SELECT anon_1.id, anon_1.id_src, anon_1.id_dst 
FROM (SELECT link.id AS id, remap_1.id_low AS id_src, remap_2.id_low AS id_dst 
FROM link JOIN remap AS remap_1 ON link.id_src = remap_1.id_high JOIN remap AS remap_2 ON link.id_dst = remap_2.id_high) AS anon_1 
WHERE anon_1.id_src = :param_1
```
and result:

`[(<FusedLink 2 / 1 -> 1>,), (<FusedLink 3 / 1 -> 4>,)]`

  
going through this issue, `with_parent()` is confusing me with its failure modes so I will look to see if it can be clearer about what's happening.

**Reply by wwoods** [reply]
Thanks! Frankly, I figured there was a solution given the flexibility of SqlAlchemy. The reason I added `remote` and `foreign` was because, without them, executing the `with_parent` clause would raise an exception. It's good to know that the error was basically a combination of A) not needing `remote()`, B) putting `foreign()` on the wrong side, and C) querying based on the wrong side of the `sa.orm.relationship`.

Perhaps the real issue here is a lack of error message that the query being specified doesn't make sense. Again, the query generated doesn't use the object passed to `with_parent` at all -- isn't that necessarily an error? Is there any case where one would want to use `with_parent` and ignore its object argument?

The other side of this is that similar documentation would be helpful. I was following https://docs.sqlalchemy.org/en/14/orm/join_conditions.html#using-custom-operators-in-join-conditions, where `foreign()` is applied to the entity type being joined in, and not the entity type to which the relationship belongs. I very well may be missing something, but I cannot find examples where `foreign()` gets specified the way you did in your response.

Either way, thanks again -- really appreciate you taking the time to diagnose and document the fix.

**Reply by zzzeek** [reply]
> Thanks! Frankly, I figured there was a solution given the flexibility of SqlAlchemy. The reason I added `remote` and `foreign` was because, without them, executing the `with_parent` clause would raise an exception.

so the first odd thing I see is why arent we raising some kind of warning or exception when remote() is put on the local side.   there might be some weird use cases where it's valid, but i will see if i can try warning for that.

> 
> Perhaps the real issue here is a lack of error message that the query being specified doesn't make sense. Again, the query generated doesn't use the object passed to `with_parent` at all -- isn't that necessarily an error? 

right so what happened here was, with_parent() was making a condition that did not derive any state from the object, and that's strange that it just did that without a problem, additionally when i was playing with this it fails un-gracefully for the case when the wrong relationship is passed for the object.    since we are trying to get people off of "dynamic", with_parent() will be getting lots more attention as was the case here.

> Is there any case where one would want to use `with_parent` and ignore its object argument?

i dont think so.


> 
> The other side of this is that similar documentation would be helpful. I was following https://docs.sqlalchemy.org/en/14/orm/join_conditions.html#using-custom-operators-in-join-conditions, where `foreign()` is applied to the entity type being joined in, and not the entity type to which the relationship belongs. I very well may be missing something, but I cannot find examples where `foreign()` gets specified the way you did in your response.

well your setup is already a lot more complicated than what we have in the docs, but it *is* what I had long ago envisioned how people would use this.   I would think if foreign() surrounds an expression that we know to be unique, that's somnething we can warn for, however it's probably not possible to know that in all cases; even if foreign(col) is around a column that has "primary_key", if it's derived from some bigger SELECT that makes it non unique, we dont really know.        this is one of those things that is easier if one is thinking in terms of relational structures, but even for me this is not always intuitive so i dont have a great idea yet on how to better explain the role of foreign().


> 
> Either way, thanks again -- really appreciate you taking the time to diagnose and document the fix.

**Comment by zzzeek** [comment]
editing, I got the last part wrong, moment
